File: apps/server/agent/src/session/projects.ts
/**
 * Project CRUD operations against the global database.
 *
 * Sessions belonging to a project live as SQLite files under the project's
 * storage directory; `deleteProject` removes that whole tree.
 */
import type { Database as DatabaseType } from "bun:sqlite";
import crypto from "node:crypto";
import fs from "node:fs";
import path from "node:path";
import type { ProjectInfo } from "@/agent/src/types/index.js";
import type { StorageState } from "./utils.js";

interface CreateProjectOptions {
  id?: string;
  name: string;
  dir: string;
}

/** Insert or upsert a project row and return the normalized `ProjectInfo`. */
export function createProject(db: DatabaseType, options: CreateProjectOptions): ProjectInfo {
  const id = options.id ?? crypto.randomUUID();
  const now = Date.now();
  const stmt = db.prepare(`
    INSERT INTO projects (id, name, dir, created_at, updated_at)
    VALUES (?, ?, ?, ?, ?)
    ON CONFLICT(dir) DO UPDATE SET
      name = excluded.name,
      updated_at = excluded.updated_at
  `);
  stmt.run(id, options.name, options.dir, now, now);

  const row = db.prepare(`SELECT * FROM projects WHERE dir = ?`).get(options.dir) as any;
  return {
    id: row.id,
    name: row.name,
    path: row.dir,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export function getProject(db: DatabaseType, projectId: string): ProjectInfo | null {
  const row = db.prepare(`SELECT * FROM projects WHERE id = ?`).get(projectId) as any;
  if (!row) return null;
  return {
    id: row.id,
    name: row.name,
    path: row.dir,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export function getProjectByDir(db: DatabaseType, dir: string): ProjectInfo | null {
  const row = db.prepare(`SELECT * FROM projects WHERE dir = ?`).get(dir) as any;
  if (!row) return null;
  return {
    id: row.id,
    name: row.name,
    path: row.dir,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  };
}

export function listProjects(db: DatabaseType): ProjectInfo[] {
  const rows = db.prepare(`SELECT * FROM projects ORDER BY updated_at DESC`).all() as any[];
  return rows.map((row) => ({
    id: row.id,
    name: row.name,
    path: row.dir,
    createdAt: row.created_at,
    updatedAt: row.updated_at,
  }));
}

/**
 * Delete a project: close every session DB connection it owns, remove the
 * project's entire storage directory, then clear the global index rows.
 */
export function deleteProject(state: StorageState, projectId: string): boolean {
  const { globalDb, sessionDbs, storageDir } = state;

  // 1. Close cached session connections for this project.
  const sessionIds = globalDb
    .prepare(`SELECT id FROM sessions WHERE project_id = ?`)
    .all(projectId) as Array<{ id: string }>;
  for (const { id } of sessionIds) {
    const db = sessionDbs.get(id);
    if (db) {
      try {
        db.close();
      } catch {
        // Ignored
      }
      sessionDbs.delete(id);
    }
  }

  // 2. Delete the project's storage directory (sessions + DBs).
  try {
    const projDir = path.join(storageDir, "projects", projectId);
    if (fs.existsSync(projDir)) {
      fs.rmSync(projDir, { recursive: true, force: true });
    }
  } catch {
    // Ignored
  }

  // 3. Clear global index rows.
  globalDb.prepare(`DELETE FROM sessions WHERE project_id = ?`).run(projectId);
  const info = globalDb.prepare(`DELETE FROM projects WHERE id = ?`).run(projectId);
  return info.changes > 0;
}

File: apps/server-go/internal/session/projects.go
// Project CRUD against the global DB. Port of agent/src/session/projects.ts.
package session

import (
	"crypto/rand"
	"database/sql"
	"encoding/hex"
	"fmt"
	"os"
)

func randomID() string {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		panic(err) // crypto/rand never fails on supported platforms
	}
	return fmt.Sprintf("%s-%s-%s-%s-%s",
		hex.EncodeToString(b[0:4]), hex.EncodeToString(b[4:6]),
		hex.EncodeToString(b[6:8]), hex.EncodeToString(b[8:10]),
		hex.EncodeToString(b[10:16]))
}

type createProjectOptions struct {
	ID   string
	Name string
	Dir  string
}

func createProject(db *sql.DB, opts createProjectOptions) (ProjectInfo, error) {
	id := opts.ID
	if id == "" {
		id = randomID()
	}
	now := nowMillis()
	_, err := db.Exec(`
		INSERT INTO projects (id, name, dir, created_at, updated_at)
		VALUES (?, ?, ?, ?, ?)
		ON CONFLICT(dir) DO UPDATE SET
			name = excluded.name,
			updated_at = excluded.updated_at
	`, id, opts.Name, opts.Dir, now, now)
	if err != nil {
		return ProjectInfo{}, err
	}
	return projectByDir(db, opts.Dir)
}

func scanProject(row interface{ Scan(...any) error }) (ProjectInfo, error) {
	var p ProjectInfo
	err := row.Scan(&p.ID, &p.Name, &p.Path, &p.CreatedAt, &p.UpdatedAt)
	return p, err
}

func projectByID(db *sql.DB, id string) (ProjectInfo, error) {
	return scanProject(db.QueryRow(
		`SELECT id, name, dir, created_at, updated_at FROM projects WHERE id = ?`, id))
}

func projectByDir(db *sql.DB, dir string) (ProjectInfo, error) {
	return scanProject(db.QueryRow(
		`SELECT id, name, dir, created_at, updated_at FROM projects WHERE dir = ?`, dir))
}

func listProjects(db *sql.DB) ([]ProjectInfo, error) {
	rows, err := db.Query(
		`SELECT id, name, dir, created_at, updated_at FROM projects ORDER BY updated_at DESC`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()
	var out []ProjectInfo
	for rows.Next() {
		p, err := scanProject(rows)
		if err != nil {
			return nil, err
		}
		out = append(out, p)
	}
	return out, rows.Err()
}

// deleteProject closes cached session connections, removes the project's
// storage tree, then clears the global index rows.
func (s *Storage) deleteProject(projectID string) (bool, error) {
	s.mu.Lock()
	rows, err := s.globalDB.Query(`SELECT id FROM sessions WHERE project_id = ?`, projectID)
	if err != nil {
		s.mu.Unlock()
		return false, err
	}
	var sessionIDs []string
	for rows.Next() {
		var id string
		if err := rows.Scan(&id); err != nil {
			rows.Close()
			s.mu.Unlock()
			return false, err
		}
		sessionIDs = append(sessionIDs, id)
	}
	err = rows.Err()
	rows.Close()
	if err != nil {
		s.mu.Unlock()
		return false, err
	}
	for _, id := range sessionIDs {
		if db, ok := s.sessionDBs[id]; ok {
			_ = db.Close()
			delete(s.sessionDBs, id)
		}
	}
	s.mu.Unlock()

	_ = os.RemoveAll(ProjectStorageDir(s.storageDir, projectID))

	if _, err := s.globalDB.Exec(`DELETE FROM sessions WHERE project_id = ?`, projectID); err != nil {
		return false, err
	}
	res, err := s.globalDB.Exec(`DELETE FROM projects WHERE id = ?`, projectID)
	if err != nil {
		return false, err
	}
	n, _ := res.RowsAffected()
	return n > 0, nil
}
